crates/runtime/src/gc/enabled.rs

```diff
@@ -333,30 +344,92 @@ impl VMExternRef {
             ptr::write(
                 extern_data_ptr,
                 VMExternData {
-                    ref_count: AtomicUsize::new(1),
+                    ref_count: Cell::new(1),
                     // Cast from `*mut T` to `*mut dyn Any` here.
                     value_ptr: SendSyncPtr::new(NonNull::new_unchecked(value_ptr.as_ptr())),
                 },
             );
 
             log::trace!("New externref data @ {:p}", extern_data_ptr);
-            VMExternRef(NonNull::new_unchecked(extern_data_ptr).into())
+            VMExternRef(VMGcRef::from_non_null(NonNull::new_unchecked(
+                extern_data_ptr as *mut u8,
+            )))
         }
     }
 
-    /// Turn this `VMExternRef` into a raw, untyped pointer.
+    /// Turn this `VMExternRef` into a its underlying GC reference.
+    ///
+    /// Unlike `into_gc_ref`, this does not consume and forget `self`. It is *not*
+    /// safe to use `from_gc_ref` on pointers returned from this method; only use
+    /// `clone_from_gc_ref`!
+    ///
+    /// Nor does this method increment the reference count. You must ensure that
+    /// `self` (or some other clone of `self`) stays alive until
+    /// `clone_from_raw` is called.
+    #[inline]
+    pub fn as_gc_ref(&self) -> &VMGcRef {
+        &self.0
+    }
+
+    /// Consume this `VMExternRef` into a raw, untyped pointer.
+    ///
+    /// # Safety
+    ///
+    /// This method forgets self, so it is possible to create a leak of the
+    /// underlying reference counted data if not used carefully.
+    ///
+    /// Use `from_raw` to recreate the `VMExternRef`.
+    pub unsafe fn into_gc_ref(self) -> VMGcRef {
+        let gc_ref = self.0;
+        std::mem::forget(self);
+        gc_ref
+    }
+
+    /// Recreate a `VMExternRef` from a pointer returned from a previous call to
+    /// `{as,into}_gc_ref`.
+    ///
+    /// # Safety
+    ///
+    /// The given `gc_ref` must point to a valid `VMExternData`.
+    ///
+    /// Unlike `clone_from_raw`, this does not increment the reference count of
+    /// the underlying data. It is not safe to continue to use the pointer
+    /// passed to this function.
+    #[inline]
+    pub unsafe fn from_gc_ref(gc_ref: VMGcRef) -> Self {
```

Comment:
Could this an `clone_from_gc_ref` assert that vmexternref is the only implemented type?